src/librustc_codegen_llvm/attributes.rs

```diff
@@ -135,11 +137,28 @@ pub fn from_fn_attrs(cx: &CodegenCx, llfn: ValueRef, id: DefId) {
         Attribute::NoAlias.apply_llfn(
             llvm::AttributePlace::ReturnValue, llfn);
     }
-    if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::UNWIND) {
-        unwind(llfn, true);
-    }
-    if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::RUSTC_ALLOCATOR_NOUNWIND) {
-        unwind(llfn, false);
+
+    let can_unwind = if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::UNWIND) {
+        Some(true)
+    } else if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::RUSTC_ALLOCATOR_NOUNWIND) {
+        Some(false)
+
+    // Perhaps questionable, but we assume that anything defined
+    // *in Rust code* may unwind. Foreign items like `extern "C" {
+    // fn foo(); }` are assumed not to unwind **unless** they have
+    // a `#[unwind]` attribute.
+    } else if !cx.tcx.is_foreign_item(id) {
```

Comment:
Is this going to cause similar issues to #48251?

Reply:
Oh nevermind, seems like we've already been doing this.

Reply:
Ah yeah this was accidentally overwriting the custom attributes we have so I just moved this around to make the attributes (unstable) take priority